README
		# -*- coding:utf-8; mode:Text; fill-column:79 -*-
		# Time-stamp: "2014-06-13 19:01:57 MDT sburke@cpan.org"
# (This page is in UTF-8!) |··················································|

Module:  Text::Unidecode-- make ASCII transliterations of Unicode text

Unidecode makes ASCII transliterations of Unicode text.  Sometimes it's dumb,
but it's better than looking at "???" or "\15BA\15A0\1610...".
If you have smarter text-handling subroutines, Unidecode might be useful
as a fallthrough for them.

Example:
   print unidecode(  "北亰\n" );
           prints:  Bei Jing

See more examples below.

For full documentation, run:
    perldoc Unidecode
Or read:see:
    http://search.cpan.org/perldoc?Text::Unidecode 
An article about how Unidecode runs:
    http://interglacial.com/tpj/22/

REQUIREMENTS
This module requires Perl 5.6.0 at the very least.  That's probably not a
problem for you, since that's from a million years ago.

INSTALLATION
* For using the "CPAN Plus" system, read:   perldoc cpanp
* For old-style "make" interface, read:     perldoc perlmodinstall

~~~ EXAMPLE UNIDECODE INPUT AND OUTPUT ~~~
(Just two or three lines, from a few languages.)

La décennie voit le début des biotechnologies avec le premier clonage,
les organismes génétiquement modifiés, le début du séquençage du
génome humain
=>  La decennie voit le debut des biotechnologies avec le premier clonage,
    les organismes genetiquement modifies, le debut du sequencage du
    genome humain

Wśród nocnej ciszy głos się rozchodzi:
Wstańcie, pasterze, Bóg się nam rodzi!
=>  Wsrod nocnej ciszy glos sie rozchodzi: 
    Wstancie, pasterze, Bog sie nam rodzi! 

Καθαίρονται δ᾽ ἄλλως αἵματι μιαινόμενοι οἷον εἴ τις εἰς πηλὸν ἐμβὰς
πηλῷ ἀπονίζοντο. μαίνεσθαι δ᾽ ἂν δοκοίη, εἴ τίς μιν ἀνθρώπων
=>  Kathairontai d' allos aimati miainomenoi oion ei tis eis pelon embas
    pelo aponizonto. mainesthai d' an dokoie, ei tis min anthropon

На другой день к завтраку подавали очень вкусные пирожки, раков и
бараньи котлеты; и пока ели, приходил наверх повар Никанор справиться,
=>  Na drughoi dien' k zavtraku podavali ochien' vkusnyie pirozhki, rakov i
    baran'i kotliety; i poka ieli, prikhodil navierkh povar Nikanor spravit'sia,

Nước trà (hay nước chè) là đồ uống phổ biến thứ hai trên thế giới (sau
nước uống). Nó làm bằng cách ngâm lá, chồi, hay cành của cây chè
=>  Nuoc tra (hay nuoc che) la do uong pho bien thu hai tren the gioi (sau
    nuoc uong). No lam bang cach ngam la, choi, hay canh cua cay che

#### And Then Things Get A Bit Suboptimal
# But remember the Unidecode motto: "It's better than nothing!"

유자차(柚子茶)는 유자청을 찬물이나 더운 물에 희석하여 마시는 한국의
전통 차이다. 유자청은 얇게 자른 유자를 꿀이나 설탕과 섞은 뒤 3~4개월
=>  yujaca(You Zi Cha )neun yujaceongeul canmulina deoun mule hyiseoghayeo masineun hangugyi
    jeontong caida. yujaceongeun yalbge jareun yujareul ggulina seoltanggwa seoggeun dwi 3~4gaeweol

* The Gayatri Mantra- Sanskrit
ॐ भूर्भुवः॒ स्वः ।  तत्स॑वितुर्वरे॑णियं ।
भ॒र्गो॑ दे॒वस्य॑ धीमहि। ।  धियो॒ यो नः॑ प्रचो॒दया॑त्॥ । 
=>  AUM bhuurbhuvH' svH  /   tts'viturvre'nniyN  / 
    bh'rgo' de'vsy' dhiimhi /   /   dhiyo' yo nH' prco'dyaa't //   /  

道可道，非常道。名可名，非常名。無名天地之始；有名萬物之母。故常無欲，
以觀其妙；常有欲，以觀其徼。此兩者，同出而異名，同謂之玄。玄之又玄，衆
=>  Dao Ke Dao ,Fei Chang Dao . Ming Ke Ming ,Fei Chang Ming . Wu Ming Tian Di Zhi Shi ;You Ming Wan Wu Zhi Mu . Gu Chang Wu Yu ,
    Yi Guan Qi Miao ;Chang You Yu ,Yi Guan Qi Jiao . Ci Liang Zhe ,Tong Chu Er Yi Ming ,Tong Wei Zhi Xuan . Xuan Zhi You Xuan ,Zhong 

#Yiddish. Directionality and ligature might come out wrong in your browser/editor:]
‏טיי איז א געטראנק וואס מען טרינקט איבער דער גארער וועלט. טיי ווערט
געמאכט דורך ווייקן די געטרוקנטע בלעטער אדער בלומען פון דעם פלאנץ‎
=>  tyy yz  g`trnq vvs m`n trynqt yb`r d`r gr`r vv`lt. tyy vv`rt
    g`mkt dvrk vvyyqn dy g`trvqnt` bl`t`r d`r blvm`n pvn d`m plnts


#Urdu. Directionality and ligature might come out wrong in your browser/editor:]
‏چائے دنیا کی پسندیدہ مشروب ہے۔ یہ چاۓ کے پودے کی پتیوں کو چند منٹ گرم
پانی میں ابالنے سے تیار ہوتی ہے۔‎
=>  chy'y dny khy psndydh mshrwb hy. yh chy' khy pwdy khy ptywN khw chnd mntt grm
    pny myN blny sy tyr hwty hy.



ChangeLog
 -*-coding:utf-8;fill-column:79-*- ············································
 Last Modified Time-stamp: "2014-12-07 05:39:39 MST sburke@cpan.org"
#==============================================================================

Revision history for Perl module Text::Unidecode

2014-12-07   Sean M. Burke  sburke@cpan.org
	* RELEASE 1.23.  Just a bugfix version.
	* The bug in question: https://rt.cpan.org/Ticket/Display.html?id=97456
	* Thank you very much to superstar Dagfinn Ilmari Mannsaker for noting
	  it first *and* for providing a patch for a problem that would baffle
	  me completely:
	   "On perls 5.8.8 through 5.12.x, regex matches against UTF-16
  	   surrogate characters emits a fatal "Malformed UTF-8 character"
	   warning if warnings are enabled. ExtUtils::MakeMaker prior to 6.78
	   runs the test suite with -w, causing the installation to fail.
	   The attached patch [which I applied -SMB] disables utf8
	   warnings while doing the regex substitution and converting the
	   character number to a character in the test."
 	  And thank you very much to Ricardo Signes and Tim Bunce for reminding
	  me to actually release this thang!  I was stupid and forgot... for
	  several MONTHS.	
	* Doc: Adding mention of Tom Christiansen's "Perl Unicode Cookbook":
	     http://www.perl.com/pub/2012/04/perlunicook-standard-preamble.html
	* Doc: Adding a suggestion of "use utf8;" in German example.
	
2014-08-15   Sean M. Burke  sburke@cpan.org
	* RELEASE 1.22.  (The dev release works, so this is a version bump.)
	* See notes for 2014-07-25, because this is the first public release
	with significant changes since 2001!
	
2014-07-25   Sean M. Burke  sburke@cpan.org
	* !DEVELOPER RELEASE!
	* !Release 1.20_01!
	* Many bugfixes.  Thanks especially to Tomaž Šolc!
	* Yet more *.t files added for improved sanity checking.
	* Shuffling around the internals of Unidecode.pm
	* Putting in some vacuous 0x__.pm files where
	previously there would just be a load failure
	
2014-06-30   Sean M. Burke  sburke@cpan.org
	* Release 1.01 -- first official Unidecode release since 2001!!!
	* There are no real changes since the 2014-06-23 developer
	release.  I'm just making this all official now.
	
	
2014-06-23   Sean M. Burke  sburke@cpan.org
	* !DEVELOPER RELEASE!
	* Release 1.00_03
	* Now asserting that we need at least Perl 5.8.0
	An automated test system that tried running the t/*.t
        under a 5.6.2 spewed all kinds of crazy error messages.
	Hence the bump-up.
	So, I added assertions for the version.
	* I added some tests for more basic sanity assertions.
	
2014-06-17   Sean M. Burke  sburke@cpan.org
	v1.00_02 - Not released.  Just internal rearranging.
	
	
2014-06-13   Sean M. Burke  sburke@cpan.org
	* !DEVELOPER RELEASE!
	* Release 1.00(_01!)- so many years later, finally we bump up to 1.*!

	* My documentation is now BRILLIANT.
	* Minor bugfixes.
	* Some code comments for clarity.
	* A modern test suite.

	* A proper release will follow in a few days.
	
	
2001-07-14  Sean M. Burke  sburke@cpan.org
	
	* Release 0.04 -- forgot to put TODO.txt in 0.03.  Now including
	it.  That's the only change.
	
2001-07-14  Sean M. Burke  sburke@cpan.org
	
	* Release 0.03 -- first public release.

[END OF CHANGELOG]


lib/Text/Unidecode/__check_all_here.pl
#!/usr/bin/perl
#: check that we have all tables here, 0x00__ to 0x_ff
#==============================================================================
#### See the end of this file for the distribution/copyright license.
# sburke@cpan.org
our $Last_Modified =' Time-stamp: "2014-07-25 02:59:20 MDT sburke@cpan.org"';
use strict; use warnings; use Carp;
our $VERSION= "0.01"; 
use constant DEBUG => 10;

use Text::Unidecode ();
$| = 1;
Main_Run();

exit;
#======================================================================

sub Main_Run {
  shift @ARGV if @ARGV and $ARGV[0] eq '--';
  die "I take no arguments" if @ARGV;

  my(@glob) = dir_files_matching("./",
    qr/    # like: "x0c.pm"
       \A
         x
         [0-9a-fA-F]{2}
         \.pm
         \z
      /x
  );

  #print "Them: @glob\n";

  foreach my $block_number ( 0x00 .. 0xFF ) {
    my $b_hex = sprintf "%02x", $block_number;
    my $fs = "x$b_hex.pm"; # like "x32.pm"
    unless(-e $fs) {
      print "Missing block: $fs\n";
      gen_block($fs, $b_hex);
    }

    my $charnum = 0xFE + 0xFF * $block_number;
    my $char = sprintf( "%c", $charnum );
    #printf( "# About to load table %s via char %02x)\n", $hex, $charnum );
    my $x = Text::Unidecode::unidecode($char);
    
  }

  print "Done.\n";

  return;
}
#======================================================================

sub gen_block {
  my($fs, $b_hex) = @_;
  die "WHERE!?" unless $fs and $fs =~ m/\S/;
  die "IT EXISTS!?!" if -e $fs;

  open my $NEW, ">", $fs or die "Can't write-open $fs - $!";
  print "  Generating $fs ...\n";
  
  my $code = 
  qq{# $Last_Modified
\$Text::Unidecode::Char[ 0x$b_hex ] = Text::Unidecode::make_placeholder_map();
1;
};

  print $NEW $code or die "Can't write to $fs - $!";
  close($NEW) or die "Can't close $fs - $!";

  return;
}

sub dir_files_matching {
  my($dirspec,$re) = @_;
  my(@items);
  {
    opendir my $D, $dirspec or die "Can't opendir $dirspec - $!";
    @items = readdir($D);
    closedir($D) or die "Can't closedir $dirspec - $!";
  }
  @items = grep $_ =~ $re, @items;
  return sort @items;
}

#======================================================================
__END__


Makefile.PL
# This -*-perl-*- script writes the Makefile for installing this distribution.
# ( Last Modified Time-stamp: "2014-06-17 19:30:16 MDT sburke@cpan.org" )
#======================================================================
#
# For info on instaling this module and how to control it, see:
#   perldoc cpan
#   perldoc cpanp
#   perldoc perlmodinstall
#   perldoc ExtUtils::MakeMaker
#
#
#======================================================================

require                   5.008_000;  #   # <==========!!!!!!!!!!
my $min_perl_version =    5.008_000;  #   # <==========!!!!!!!!!!

# For the module author: see lib/ExtUtils/MakeMaker.pm for details
#   of how to influence the contents of the Makefile that is written.
# Incidentally. Perldoc perlfunc says "use " is compile-time, but
#   "require 5.6" is runtime.

use ExtUtils::MakeMaker;
#if( $ExtUtils::MakeMaker::VERSION < 6.47 ) {
#  die "I need at least perl version $min_perl_version";
#}

WriteMakefile(
  'NAME'         => 'Text::Unidecode',
  'ABSTRACT'     => 'Provide plain ASCII transliterations of Unicode text',

  'VERSION_FROM' => 'lib/Text/Unidecode.pm', # finds $VERSION

  'AUTHOR'       => 'Sean Burke <sburke@cpan.org>',
  'LICENSE'      => 'perl',
                   #  "...under the same terms as Perl itself"

  ($ExtUtils::MakeMaker::VERSION >= 6.47
    ? ('MIN_PERL_VERSION' => $min_perl_version )

    : () # Otherwise we can't express that metadata about this
         #  whole module suite, and we just have to let the version
         #  "require" thing up there block this whole Makefile.PL
         #  from running.
  ),

  'dist'  => {
     'COMPRESS' => 'gzip -4f',
       # from 4 to 6 is where there's a sudden doubling
       #  of processing time,  So 4 is cool.
     'SUFFIX'   => 'gz',
     'TARFLAGS' => 'cfh', # h=follow symlinks when making dist
  },

);

__END__
